lib/core/src/Cardano/Wallet.hs

```diff
@@ -686,29 +693,46 @@ newWalletLayer tracer bp db nw tl = do
             -- block of the list, even if empty, so that we correctly update the
             -- current tip of the wallet state.
-            let nonEmpty = not . null . transactions
-            let (h, q) = first (filter nonEmpty) $
-                    NE.splitAt (length blocks - 1) blocks
-            liftIO $ logDebug t $ pretty (h ++ q)
-            let (txs, cp') = NE.last $ applyBlocks @s @t (h ++ q) cp
-            let progress = slotRatio epochLength sup tip
-            let status' = if progress == maxBound
-                    then Ready
-                    else Restoring progress
-            let meta' = meta { status = status' } :: WalletMetadata
-            let nPending = Set.size (getPending cp')
-            liftIO $ logInfo t $ pretty meta'
-            liftIO $ logInfo t $ nPending ||+" transaction(s) pending."
-            liftIO $ logInfo t $
-                length txs ||+ " new transaction(s) discovered."
-            let (Quantity bh) = blockHeight cp'
-            liftIO $ logInfo t $
-                "block height is "+||bh||+""
-            unless (null txs) $ liftIO $ logDebug t $
-                pretty $ blockListF (snd <$> Map.elems txs)
-            DB.putCheckpoint db (PrimaryKey wid) cp'
-            DB.putTxHistory db (PrimaryKey wid) txs
-            DB.putWalletMeta db (PrimaryKey wid) meta'
+            let nonEmptyBlocks = biconcat
+                    $ first (filter $ not . null . transactions)
+                    $ NE.splitAt (length blocks - 1) blocks
+
+            let calculateMetadata :: SlotId -> WalletMetadata
+                calculateMetadata slot = meta { status = status' }
+                  where
+                    progress' = slotRatio epochLength slot nodeTip
+                    status' =
+                        if progress' == maxBound
+                        then Ready
+                        else Restoring progress'
+
+            let txs_cps = applyBlocks @s @t nonEmptyBlocks cp
+            let txs = fold $ fst <$> txs_cps
+
+            let cpLast = fromMaybe cp $ safeLast $ snd <$> txs_cps
+            let (Quantity bhLast) = blockHeight cpLast
+
+            liftIO $ do
+                logDebug t $ "blocks: "
+                    <> pretty nonEmptyBlocks
+                logDebug t $ "transactions: "
+                    <> pretty (blockListF (snd <$> Map.elems txs))
+                logInfo t $ "metadata: "
+                    +|| pretty (calculateMetadata slotLast)
+                logInfo t $ "number of pending transactions: "
+                    +|| Set.size (getPending cpLast) ||+ ""
+                logInfo t $ "number of new transactions: "
+                    +|| length txs ||+ ""
+                logInfo t $ "new block height: "
+                    +|| bhLast ||+ ""
+
+            forM_ txs_cps $ \(txs', cp') -> do
+
+                let meta' = calculateMetadata (view #slotId $ currentTip cp')
```

Comment:
Since we're lensing I think that could be `currentTip cp' ^. #slotId`

Reply:
I'm happy to change it, though this module makes heavy use of `view` instead of `^.`. Perhaps we can make a later PR to change this across the board?

Reply:
OK, well if the module is not using `^.` then it's good to match that.
The line just looked strange to me because `view #slotId = slotId`.